### ISSUE TITLE
Training hangs when performing an assign op on aggregated gradients in SyncReplicasOptimizer

### ISSUE THREAD

**Issue by sahiltyagi4** [tensorflow/tensorflow#41430]
I'm trying to fetch the aggregated gradients which are computed after aggregating all the replicas together in SyncReplicasOptimizer. To do this, I added an assign op on a variable that I created and assigned the aggregated gradients value (from the aggregated_grads_and_vars). Basically in sync_replicas_optimizer.py apply_gradients() call I did something like:

```
var1 = tf.Variable(0.0, trainable=False)
def apply_gradients(grads_and_vars, global_step):
   
   *aggregated_grads_and_vars returned based on replicas_to_aggregate*
   aggregated_grads_and_vars = zip(aggregated_grad, var_list)
   
   test_var = tf.assign(var1, aggregated_gradients)

   *execute everything else in the fn and return train_op*
```

Later, I call *session.run(train_op)* to execute a training steps followed by  *session.run(test_var)*.
My training process goes through with the first training and then hangs. The problem is with *session.run(test_var)*. 

My approach works when I use any other asynchronous optimizer (tested on *MomentumOptimizer*, *AdamOptimizer* and *GradientDescentOptimizer*) but not for *SyncReplicasOptimizer*.

Could this have something to do with *SyncReplicasOptimizerHook* ? 

Any lead would be greatly appreciated.

**Comment by amahendrakar** [comment]
@sahiltyagi4,
In order to expedite the trouble-shooting process, could you please provide the complete code to reproduce the issue reported here and also the TensorFlow version you are using. Thanks!

**Comment by sahiltyagi4** [comment]
I'm using Tensorflow r1.14.
The source code is available at [https://github.com/sahiltyagi4/tensorflow/tree/r1.14](url).

In particular, I made changes to apply_gradients fn in sync_replicas_optimizer.py at lines [https://github.com/sahiltyagi4/tensorflow/blob/r1.14/tensorflow/python/training/sync_replicas_optimizer.py#L285](url), and doing the assign op the values at [https://github.com/sahiltyagi4/tensorflow/blob/r1.14/tensorflow/python/training/sync_replicas_optimizer.py#L285](url).

I build this tensorflow (WITHOUT CUDA, MPI, etc.) and use this in Resnet at [https://github.com/sahiltyagi4/tensorflow/blob/r1.14/models/official/resnet/resnet_regular.py](url).

When I call *mon_sess.run()* assign op in *sync_replicas_optimizer.py line 285* like I have in [https://github.com/sahiltyagi4/estimator/blob/dynamic_v2/tensorflow_estimator/python/estimator/estimator.py#L1542](url) (I build the estimator from source too btw), that's when my training comes to a halt (know this because previous *mon_sess.run* on train_op executes for the first step).

Commenting out the assign op, my training executes as it should. Also, in  the apply_gradients fn, when I use the grad_and_vars (returned from compute_gradients fn) instead of aggregated_grads_and_vars (from SyncReplicasOptimizer), I'm able to execute the assign op and fetch its value successfully.

Would also like to point out that when I did the same changes in any other optimizer (in apply_gradients fn of optimizer.py and don't wrap it with SyncReplicasOptimizer), then this works. So I think there's something happening with the *SyncReplicasOptimizerHook* maybe?

**Comment by sahiltyagi4** [comment]
@amahendrakar were you able to see the problem?

**Comment by amahendrakar** [comment]
@sahiltyagi4,
Thank you for the update. The example is fairly complex, and it is hard for us to pinpoint the issue.

Can you please get the example down to the simplest possible repro? That will allow to determine the source of the issue easily. Thanks!

**Comment by amahendrakar** [comment]
Also, is there a specific reason you are using TF v1.14? Please update TensorFlow to v2.2 or v1.15 and check if you are facing the same issue. Thanks!
